### ISSUE TITLE
How to save the image shown by shap?

### ISSUE THREAD

**Issue by RoachLin** [shap/shap#1243]
https://github.com/slundberg/shap/blob/master/notebooks/deep_explainer/PyTorch%20Deep%20Explainer%20MNIST%20example.ipynb

could you tell me how to save the image shown by "shap.image_plot(shap_numpy, -test_numpy)" ?

**Comment by giangnguyen2412** [comment]
Did you know how? I am facing the same problem. Thanks!

**Comment by RoachLin** [comment]
> Did you know how? I am facing the same problem. Thanks!

Sorry, I don't know how. I just use a software to make a screenshot.

**Comment by giangnguyen2412** [comment]
when I run `shap.image_plot(shap_numpy, -test_numpy)`, the input image is all black. Do you know why?

**Comment by RoachLin** [comment]
> when I run `shap.image_plot(shap_numpy, -test_numpy)`, the input image is all black. Do you know why?

I made some changes to the example code to make it work. It's long time ago so I can't remember it exactly. Maybe the problem is using cpu or gpu. When using gpu, the output image should transform to cuda. I will show my code below:
```
# PyTorch Deep Explainer MNIST example
# A simple example showing how to explain an MNIST CNN trained using PyTorch with Deep Explainer.

import numpy as np
import shap
import time
import torch
from torch import nn, optim
from torch.nn import functional as F
from torchvision import datasets, transforms


# torch.nn是专门为神经网络设计的模块化接口。nn构建于autograd之上，可以用来定义和运行神经网络。
# nn.Module是nn中十分重要的类，包含网络各层的定义及forward方法。

# 定义自己的神经网络，需要继承nn.Module类，并实现forward方法。
class Net(nn.Module):
    # nn.Module的子类必须在构造函数中执行父类的构造函数
    def __init__(self):
        super(Net, self).__init__()  # 等价于nn.Module.__init__()

        self.conv_layers = nn.Sequential(
            nn.Conv2d(1, 10, kernel_size=5),
            nn.MaxPool2d(2),
            nn.ReLU(),
            nn.Conv2d(10, 20, kernel_size=5),
            nn.Dropout(),
            nn.MaxPool2d(2),
            nn.ReLU())

        self.fc_layers = nn.Sequential(
            nn.Linear(320, 50),
            nn.ReLU(),
            nn.Dropout(),
            nn.Linear(50, 10),
            nn.Softmax(dim=1))

    # 只要在nn.Module的子类中定义了forward函数，backward函数就会被自动实现（利用autograd）。
    def forward(self, x):
        x = self.fc_layers(self.conv_layers(x).view(-1, 320))
        return x


def train(model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output.log(), target)
        loss.backward()
        optimizer.step()
        if batch_idx % 100 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch,
                batch_idx * len(data),
                len(train_loader.dataset),
                100. * batch_idx / len(train_loader),
                loss.item()))


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output.log(), target).item()  # sum up batch loss
            pred = output.max(1, keepdim=True)[1]  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss,
        correct,
        len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


# 我们可以一次性将整个数据集喂给神经网络，让神经网络利用全部样本来计算迭代时的梯度（即传统的梯度下降法），
# 也可以一次只喂一个样本（即随机梯度下降法，也称在线梯度下降法），
# 也可以取个折中的方案，即每次喂一部分样本让其完成本轮迭代（即batch梯度下降法）。
batch_size = 128  # 一次训练所选取的样本数
num_epochs = 10  # 1个epoch表示过了1遍训练集中的所有样本
# device = torch.device('cpu')
device = torch.device('cuda')  # cuda指使用GPU加速，可选参数还有上面的cpu
torch.backends.cudnn.benchmark = True  # 如果上一行的参数是cpu，那这行就注释掉
# 设置torch.backends.cudnn.benchmark = True将会让程序在开始时花费一点额外时间，
# 为整个网络的每个卷积层搜索最适合它的卷积实现算法，进而实现网络的加速。
# 适用场景是网络结构固定（不是动态变化的），网络的输入形状（包括batch_size，图片大小，输入的通道）是不变的，
# 其实也就是一般情况下都比较适用。反之，如果卷积层的设置一直变化，将会导致程序不停地做优化，反而会耗费更多的时间。

train_loader = torch.utils.data.DataLoader(
    datasets.MNIST(
        'mnist_data',
        train=True,
        download=True,
        transform=transforms.Compose([transforms.ToTensor()])),
    batch_size=batch_size,
    shuffle=True)

test_loader = torch.utils.data.DataLoader(
    datasets.MNIST(
        'mnist_data',
        train=False,
        transform=transforms.Compose([transforms.ToTensor()])),
    batch_size=batch_size,
    shuffle=True)

model = Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.5)

start = time.perf_counter()

for epoch in range(1, num_epochs + 1):
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

end = time.perf_counter()
print((end - start), "s")

# 画图
# since shuffle=True, this is a random sample of test data
images, _ = next(iter(test_loader))
background = images[:100].cuda()  # 上面处理时转为cpu tensor了，所以要转回cuda
test_images = images[100:105]
shap_values = shap.DeepExplainer(model, background).shap_values(test_images)
shap_numpy = [np.swapaxes(np.swapaxes(s, 1, -1), 1, 2) for s in shap_values]
test_numpy = np.swapaxes(np.swapaxes(test_images.numpy(), 1, -1), 1, 2)
shap.image_plot(shap_numpy, -test_numpy)  # plot the feature attributions
```

**Comment by giangnguyen2412** [comment]
oh, I am running on ImageNet data so I think i can not use this code. Anyway, it should be useful for someone. Thanks for sharing.

**Comment by giangnguyen2412** [comment]
> when I run `shap.image_plot(shap_numpy, -test_numpy)`, the input image is all black. Do you know why?
I found the solution. Because the value of test_numpy is too small compared to shap_numpy values. I normalized two variables to the same scale and it works. Note that I am running on Imagenet data, not mnist
